File: ios-localwave/LocalWave/AppState/PeopleViewModel.swift
import Foundation

@MainActor
public final class PeopleViewModel: ObservableObject {
    @Published public private(set) var peers: [PeerProfile] = []
    @Published public private(set) var transportState = TransportState()
    @Published public private(set) var wakeStateByPeer: [PeerID: WakeButtonState] = [:]
    @Published public private(set) var wakeError: String?

    public let environment: AppEnvironment
    public let channel: ChannelCode
    public let displayName: String

    private var peerTask: Task<Void, Never>?
    private var transportTask: Task<Void, Never>?

    public init(environment: AppEnvironment, channel: ChannelCode, displayName: String) {
        self.environment = environment
        self.channel = channel
        self.displayName = displayName
    }

    deinit {
        peerTask?.cancel()
        transportTask?.cancel()
    }

    public var scanningStatusText: String {
        switch transportState.permission {
        case .denied, .unavailable:
            return "Bluetooth permission needed"
        case .unknown:
            return "Checking Bluetooth..."
        case .allowed:
            if transportState.isScanning {
                return "Scanning nearby..."
            }
            return peers.isEmpty ? "No one nearby yet." : "Offline local mode active"
        }
    }

    public var needsBluetoothBanner: Bool {
        transportState.permission == .denied || transportState.permission == .unavailable || transportState.permission == .unknown
    }

    public func start() {
        guard peerTask == nil, transportTask == nil else { return }

        peerTask = Task { [environment] in
            for await peers in environment.engine.observePeers() {
                if Task.isCancelled { return }
                self.peers = peers.sorted { lhs, rhs in
                    if lhs.state == rhs.state {
                        return lhs.displayName.localizedCaseInsensitiveCompare(rhs.displayName) == .orderedAscending
                    }
                    return lhs.state.sortOrder < rhs.state.sortOrder
                }
            }
        }

        transportTask = Task { [environment] in
            for await state in environment.engine.observeTransportState() {
                if Task.isCancelled { return }
                self.transportState = state
            }
        }
    }

    public func stop() {
        peerTask?.cancel()
        transportTask?.cancel()
        peerTask = nil
        transportTask = nil
    }

    public func wakeState(for peer: PeerProfile) -> WakeButtonState {
        if peer.state != .available && peer.state != .connecting {
            return .unavailable
        }
        if transportState.permission != .allowed {
            return .permissionNeeded
        }
        return wakeStateByPeer[peer.id] ?? .idle
    }

    public func sendWake(to peer: PeerProfile) async {
        let state = wakeState(for: peer)
        guard state == .idle || state == .failed else { return }

        wakeError = nil
        wakeStateByPeer[peer.id] = .sending
        do {
            try await environment.engine.sendWake(to: peer.id)
            wakeStateByPeer[peer.id] = .sent
            try? await Task.sleep(nanoseconds: 1_200_000_000)
            if !Task.isCancelled, wakeStateByPeer[peer.id] == .sent {
                wakeStateByPeer[peer.id] = .idle
            }
        } catch {
            wakeStateByPeer[peer.id] = .failed
            wakeError = "Wake could not be sent. This person may not be reachable right now."
        }
    }
}

private extension PresenceState {
    var sortOrder: Int {
        switch self {
        case .available: return 0
        case .connecting: return 1
        case .recentlySeen: return 2
        case .sleeping: return 3
        case .permissionNeeded: return 4
        }
    }
}

File: android-localwave/app/src/main/java/com/localwave/feature/people/PeopleViewModel.kt
package com.localwave.feature.people

import com.localwave.core.model.PeerId
import com.localwave.core.model.PeerProfile
import com.localwave.core.model.PresenceState
import com.localwave.core.model.TransportPermissionState
import com.localwave.core.model.TransportState
import com.localwave.core.model.WakeButtonState
import com.localwave.core.protocol.LocalWaveEngine
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class PeopleUiState(
    val peers: List<PeerProfile> = emptyList(),
    val transportState: TransportState = TransportState(),
    val wakeStates: Map<PeerId, WakeButtonState> = emptyMap(),
    val error: String? = null
)

class PeopleViewModel(private val engine: LocalWaveEngine) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private val _state = MutableStateFlow(PeopleUiState())
    val state: StateFlow<PeopleUiState> = _state

    fun start() {
        scope.launch {
            engine.observePeers().collectLatest { peers ->
                _state.update { it.copy(peers = peers.sortedBy { peer -> peer.state.sortOrder }) }
            }
        }
        scope.launch {
            engine.observeTransportState().collectLatest { transport ->
                _state.update { it.copy(transportState = transport) }
            }
        }
    }

    fun wakeState(peer: PeerProfile): WakeButtonState {
        val transport = state.value.transportState
        if (transport.permission != TransportPermissionState.ALLOWED) return WakeButtonState.PERMISSION_NEEDED
        if (peer.state !in setOf(PresenceState.AVAILABLE, PresenceState.CONNECTING)) return WakeButtonState.UNAVAILABLE
        return state.value.wakeStates[peer.id] ?: WakeButtonState.IDLE
    }

    fun sendWake(peer: PeerProfile) {
        if (wakeState(peer) !in setOf(WakeButtonState.IDLE, WakeButtonState.FAILED)) return
        _state.update { it.copy(wakeStates = it.wakeStates + (peer.id to WakeButtonState.SENDING), error = null) }
        scope.launch {
            try {
                engine.sendWake(peer.id)
                _state.update { it.copy(wakeStates = it.wakeStates + (peer.id to WakeButtonState.SENT)) }
            } catch (_: Throwable) {
                _state.update { it.copy(wakeStates = it.wakeStates + (peer.id to WakeButtonState.FAILED), error = "Wake could not be sent. This person may not be reachable right now.") }
            }
        }
    }

    private val PresenceState.sortOrder: Int get() = when (this) {
        PresenceState.AVAILABLE -> 0
        PresenceState.CONNECTING -> 1
        PresenceState.RECENTLY_SEEN -> 2
        PresenceState.SLEEPING -> 3
        PresenceState.PERMISSION_NEEDED -> 4
        PresenceState.UNREACHABLE -> 5
    }
}
